touch broadcast when updating associated schedule

domangi opened the pull request:
When a new station is added to an existing broadcast, then the updated_at field on the broadcast object should be updated. This is done by adding `touch: true` in the `belongs_to :broadcast` association of schedule.rb

roschaefer commented on features/broadcast/edit/add_many_stations.feature:
```diff
@@ -31,3 +31,11 @@ Feature: Add many stations
+    When I add "WDR Fernsehen" to the list of stations
+    And I click on "Update"
+    And I visit the broadcast page for "Die Sendung mit der Maus"
+    Then I can see "Last updated at" TODAY
```
Do you know [TimeCop](https://github.com/travisjeffery/timecop)? I would suggest to use that and say `given today is Friday 13th 2017` in the feature. By that, you can expect literal datetimes which is a good strategy to avoid bugs in the test setup.

domangi replied:
Yes I noticed that the problem was the way I was setting the current date. I will fix it with time travel as you suggested. Thanks!!

roschaefer commented on backend/app/models/schedule.rb:
```diff
@@ -1,6 +1,6 @@
 class Schedule < ApplicationRecord
   has_paper_trail
-  belongs_to :broadcast
+  belongs_to :broadcast, touch: true
```
excellent

domangi commented:
I added time travel also to other to features that were failing now that the updated_at field of broadcast is updated after an associated station is created.
I know that it is not the correct branch for it, but I also pushed a fix for the svg feature. I removed some attributes that made the comparison fail, e.g. width, height, x, y, ...

roschaefer commented on features/step_definitions/steps.rb:
```diff
@@ -757,6 +757,12 @@ def normalize_highcharts_svg(content)
   doc.css('*').remove_attr('style')
   doc.css('*').remove_attr('id')
   doc.css('*').remove_attr('clip-path')
+  doc.css('*').remove_attr('width')
```
How does the svg look like on your machine? I would remove all the attributes *except* geometries, like height and width etc. To make this step definition more robust against new versions of `highcharts`.

Does the downloaded svg look OK?

roschaefer commented on features/support/hooks.rb:
```diff
@@ -0,0 +1,4 @@
+After('@time-travel') do |scenario, block|
+  Timecop.return
+  puts "Welcome back in #{Time.now}"
```
I would remove debug output. And you can `Timecop.return` from every scenario if you wish, Timecop does not slow down the test suite significantly.

roschaefer commented on features/broadcast/edit/add_many_stations.feature:
```diff
@@ -31,3 +32,15 @@ Feature: Add many stations
+    And the current date is now
+    And I click on "Update"
+    And I am on the broadcast page for "Tom und Jerry"
+    Then I can see Last updated at is not "02/02/2017"
```
In order to avoid false negatives, I suggest to avoid `expect.not_to eq ...`.

My apologies for examples in the current code base, where we use `not_to eq` etc. It's better to use specific assertions of literal values. Expecting everything except one value can lead to errors getting through.

domangi replied:
I know, and I wanted to check the exact date using Time travel, but then I realized that it does not work  because in this case the date field is set by a different app (backend) so time travel has no effect, and I didn't want to compare the date with Time.now, but perhaps it is better than just checking a change of date.

roschaefer commented on features/broadcast/edit/add_many_stations.feature:
```diff
@@ -31,3 +32,15 @@ Feature: Add many stations
+    Given the current date is "2017-02-02"
+    And we have these broadcasts in our database:
+      | Title                    | Station   | Medium | Created at       | Updated at       |
+      | Tom und Jerry            | Das Erste | TV     | 2017-02-02 15:00 | 2017-02-02 15:00 |
```
In the step definition of `we have these broadcasts in our database:` check if the `Updated at` attribute was given. Make sure the `updated_at` of the broadcast is set to the given value, if necessary by force (`update_column`).

domangi replied:
I used the step before to set the datetime with timecop. But you are right, this should be indipendent of other steps. If the spec specifies updated_at then the step should return that, regardless of the current time.

roschaefer commented on features/broadcast/edit/add_many_stations.feature:
```diff
@@ -31,3 +32,15 @@ Feature: Add many stations
+      | Tom und Jerry            | Das Erste | TV     | 2017-02-02 15:00 | 2017-02-02 15:00 |
+    And I am on the edit page for Broadcast "Tom und Jerry"
+    When I add "WDR Fernsehen" to the list of stations
+    And the current date is now
```
Either reuse `and the current date is "2017-02-04"` or say `and I come back two days later` which reads a little nicer.

domangi replied:
Yes, it does not sound very well. I come back (two days later)? sound really better

roschaefer commented on features/step_definitions/steps.rb:
```diff
@@ -957,3 +963,31 @@ def normalize_highcharts_svg(content)
+
+When("I am on the broadcast page for {string}") do |title|
+  broadcast = Broadcast.find_by(title: title)
+  visit "/broadcast/#{broadcast.id}"
```
Really just a nitpicking, but you can use `broadcast_path(broadcast)` here :wink:

roschaefer replied:
no you can't :see_no_evil: it's Ember not Rails!

domangi replied:
I tried using paths before realizing that it was not rails 😆

roschaefer commented on features/step_definitions/steps.rb:
```diff
@@ -957,3 +963,31 @@ def normalize_highcharts_svg(content)
+end
+
+Given("the current date is {string}") do |date_string|
+  Timecop.travel(Time.parse(date_string + " 15:00 UTC +00:00"))
```
You can just `Time.parse(date_string)`.

It is pretty irrelevant for this test case, but maybe you want to develop the habit of using `Time.zone.parse`. See here: https://robots.thoughtbot.com/its-about-time-zones

domangi replied:
The problem is that when I see Time.parse("02-02-2017") let's say, it was using "2017-02-02 +01:00" and when I go to check the date I get back "2016-02-01 23:00 UTC". So check about dates failed since the date was different. I have to try to solve it directly using Time.zone, would be cleaner.
